File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/tokengen/VirologyProcessorExports.java
package uk.nhs.nhsx.virology.tokengen;

import org.apache.commons.codec.Charsets;
import uk.nhs.nhsx.virology.CtaToken;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.List;
import java.util.zip.ZipOutputStream;

import static java.util.stream.Collectors.joining;
import static uk.nhs.nhsx.core.UncheckedException.uncheckedGet;
import static uk.nhs.nhsx.core.UncheckedRun.uncheckedRun;
import static uk.nhs.nhsx.diagnosiskeydist.keydistribution.KeyFileUtility.addFileToZip;
import static uk.nhs.nhsx.diagnosiskeydist.keydistribution.KeyFileUtility.writeToFile;

public class VirologyProcessorExports {

    public static String csvFrom(List<CtaToken> tokens, String testResult, String testEndDate) {
        return tokens.stream()
            .map(it -> it.value + ", " + testResult + ", " + testEndDate)
            .collect(joining("\n", "cta token, test result, test end date\n", ""));
    }

    public static File zipFrom(String filename, CtaTokensCsv ctaTokensCsv) {
        var csvFile = uncheckedGet(() -> File.createTempFile(filename, ".csv"));
        uncheckedRun(() -> writeToFile(csvFile, ctaTokensCsv.content.getBytes(Charsets.UTF_8)));
        return uncheckedGet(() -> zipFile(filename, filename + ".csv", csvFile));
    }

    private static File zipFile(String zipFileName, String zipEntryFileName, File zipEntryFile) throws IOException {
        var zipFile = uncheckedGet(() -> File.createTempFile(zipFileName, ".zip"));
        try (var outputStream = new FileOutputStream(zipFile)) {
            try (var zipOut = new ZipOutputStream(outputStream)) {
                addFileToZip(zipOut, zipEntryFileName, zipEntryFile);
            }
        }
        return zipFile;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/tokengen/VirologyProcessorExports.kt
package uk.nhs.nhsx.virology.tokengen

import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.model.ZipParameters
import net.lingala.zip4j.model.enums.EncryptionMethod
import uk.nhs.nhsx.diagnosiskeydist.keydistribution.KeyFileUtility
import uk.nhs.nhsx.virology.CtaToken
import uk.nhs.nhsx.virology.result.TestEndDate
import uk.nhs.nhsx.virology.result.TestResult
import java.io.File

class VirologyProcessorExports(private var tempDirPath: String = System.getProperty("java.io.tmpdir") + File.separator) {

    fun csvFrom(tokens: List<CtaToken>, testResult: TestResult, testEndDate: TestEndDate): String = tokens
        .map { """${it.value}, ${testResult.wireValue}, ${TestEndDate.show(testEndDate)}""" }
        .joinToString(prefix = "cta token, test result, test end date\n", separator = "\n") { it }

    fun zipFrom(zipFileName: String, ctaTokensCsv: List<CtaTokensCsv>, zipFilePassword: String? = null): File {
        val zipFileEntries = ctaTokensCsv.map {
            val csvFile = File.createTempFile(it.filename, ".csv")
            KeyFileUtility.writeToFile(csvFile, it.content.toByteArray())
            ZipFileEntry(csvFile, it.filename)
        }

        return zipFile(zipFileName, zipFileEntries, zipFilePassword)
    }

    private fun zipFile(zipFileName: String, zipEntryFile: List<ZipFileEntry>, password: String?): File {
        val filename = "$tempDirPath$zipFileName.zip"

        val zipFile = when {
            password != null -> ZipFile(filename, password.toCharArray())
            else -> ZipFile(filename)
        }

        zipEntryFile.forEach {
            zipFile.addFile(it.file, zipParametersFrom(it, password))
        }

        return zipFile.file
    }

    private fun zipParametersFrom(it: ZipFileEntry, password: String?) = ZipParameters().apply {
        fileNameInZip = it.filename
        if (password != null) {
            isEncryptFiles = true
            encryptionMethod = EncryptionMethod.AES
        }
    }
}

data class ZipFileEntry(val file: File, val filename: String)
